### ISSUE TITLE
TODO for December 2022 Release

### ISSUE THREAD

**Issue by lbianchi-lbl** [gmlc-dispatches/dispatches#132]
## Summary

There are two distinct sets of changes that should be merged into IDAES/idaes-pse for DISPATCHES, which I'm calling very arbitrarily **A** and **B**:

### A

- **What it is** Updates to `MultiperiodModel`
- **What is affected** Needed for the nuclear case study (Although all three case studies use `MultiPeriodModel`, the updates are needed only for the nuclear case study. The updates are backwards compatible, so no changes are needed to the other case studies.)
- **How it is now**: `MultiperiodModel` is re-implemented in DISPATCHES (i.e. `import`ed from https://github.com/gmlc-dispatches/dispatches/blob/main/dispatches/models/nuclear_case/flowsheets/multiperiod.py)
- **How it should be**: `MultiperiodModel` is imported from `idaes.apps.grid_integration` (Both renewables and fossil case studies are already doing this. After the updates are merged, the nuclear case study should be able to do it too.)

### B

- **What it is** Real-time bidding
- **What is affected**
  - Currently needed by the RE case study
- **How it is now**
  - The required changes are implemented in a separate fork and branch/tag of IDAES ([dguittet:2.0.0.0a4](https://github.com/dguittet/idaes-pse/tree/2.0.0.a4))
    - Most of the changes were authored by @xiangao1; @dguittet created the branch/tag to store the changes to be able to include the RE case study in the June 2022 release during #124
    - **IMPORTANT** these changes are NOT compatible with the rest of the codebase (i.e. just setting the `idaes-pse` requirement to this ref in `setup.py` will cause the non-RE case studies to fail)
    - This link shows the differences between `dguittet:2.0.0.a4` and `IDAES:main`, i.e. "if a PR was opened from `dguittet:2.0.0.a4` to `IDAES:main`, what would it look like?"): https://github.com/IDAES/idaes-pse/compare/IDAES:main...dguittet:2.0.0.a4
  - An explicit check in various parts of the RE case study codebase ensures that the required functionality is available by looking at the version of the installed
  - To be able to test the RE notebooks in CI, a fairly clunky workaround is implemented in the CI workflow to run the RE notebooks only after re-installing `idaes-pse` from `dguittet:2.0.0.a4`
- **How it should be**: These changes should be merged into IDAES/idaes-pse together with **A** so that the same version of `idaes-pse` works for the entire DISPATCHES codebase

### C

- `ThermalGeneratorModelData` is needed by the RE case study, but missing from the current IDAES
- Fix AGC bounds in grid integration code for thermal generators
- Update NE and FE double loop flow sheets to use above updates

### D

- Fix issues with double-loop initial status (IDAES/idaes-pse#1002)

 ## PRs in `IDAES/idaes-pse`

- [x] (**A**) IDAES/idaes-pse#899
- [x] (**A**) IDAES/idaes-pse#908 
- [x] (**B**) IDAES/idaes-pse#855
- [x] (**B**) IDAES/idaes-pse#870
- [x] (**C**) IDAES/idaes-pse#986
- [x] (**C**) IDAES/idaes-pse#996
- [x] (**D**) IDAES/idaes-pse#1002

## TODO in DISPATCHES

- [x] Make all the scripts and notebooks in the repo compatible with IDAES 2.0 and switch back to upstream IDAES in `setup.py` (addressed in #151). Note: This PR will also include the changes in #139, so #139 can be closed after #151 is merged.
- [x] (**A**) Update the nuclear case study notebooks with the IDAES multiperiod import and delete `multiperiod.py` in the NE case study folder #112 (addressed in #143)
- [x] (**B**) Modify RE case study as needed (#154)
- [x] Fix convergence issues and test failures in the fossil case study (#153) (Addressed in #156) and #161
- [x] Add concrete TES unit model to the repo (addressed in #100)
- [x] Add RAVEN and TEAL dependencies to DISPATCHES (#76)

## Moved to #123

- [ ] Add steady state surrogate code from Jordan's fork (addressed in #142)
- [ ] Modify NE case study Jupyter notebook to reflect RAVEN and TEAL integration (addressed in #150)
- [ ] (**D**) Any changes connected with IDAES/idaes-pse#1002

**Comment by radhakrishnatg** [comment]
@dguittet From what I understand, you need this branch https://github.com/dguittet/idaes-pse/tree/2.0.0.a4 for the renewable case study. So, are you planning to create a PR in IDAES with these changes? If yes, it would be better to merge these changes as soon as possible.

**Comment by radhakrishnatg** [comment]
@dguittet @xiangao1 Another question. Are the changes to `Bidder`, `Coordinator` and `Tracker` classes in this branch (https://github.com/dguittet/idaes-pse/tree/2.0.0.a4) valid in general, or are they RE case study specific? @lbianchi-lbl and I were thinking we could do the following. Merge the changes to the above classes that are not case study specific with IDAES main. And, the changes that are case study specific can be included in DISPATCHES. Let us know what your thoughts are regarding this. Thank you!

**Comment by adowling2** [comment]
My vote is to generalize and contribute back to IDAES to the extent possible.

**Comment by ksbeattie** [comment]
@dguittet : We need to determine if this is generalizable and contribute this back to IDAES - which would need to happen for the IDAES Aug release.

**Comment by lbianchi-lbl** [comment]
@ksbeattie I've just updated the issue description with a more detailed description of (what I understand being) the situation. @dguittet @radhakrishnatg, could you verify that the information is correct, and add fixes and/or clarifications as needed?

**Comment by radhakrishnatg** [comment]
Sorry, @lbianchi-lbl ! I must have missed the notification, and I noticed it just now. Thanks for capturing everything we discussed! Everything looks good. I only made minor changes the description.

**Comment by dguittet** [comment]
Looks great @lbianchi-lbl 

One thing that I haven't seen written here that we've discussed is: in (B), needing to update the nuclear and fossil double loop notebooks to use the changes. Once idaes-pse has been released with the changes in (B), those notebooks released in the current dispatches are going to be incompatible with this newest idaes-pse. The next dispatches release that follows this idaes Aug release should contain changes to all 3 double loop notebooks.
